src/cell_list.h

```diff
@@ -0,0 +1,223 @@
+
+    private:
+        int calc_distance() {
+            int distance = range->distance;
+            if( min_bounds.x >= 0 ) {
+                distance = std::min( distance, range->origin.x - min_bounds.x );
+            }
+            if( min_bounds.y >= 0 ) {
+                distance = std::min( distance, range->origin.y - min_bounds.y );
+            }
+            if( max_bounds.x < MAXX ) {
+                distance = std::min( distance, max_bounds.x - range->origin.x );
+            }
+            if( max_bounds.y < MAXY ) {
+                distance = std::min( distance, max_bounds.y - range->origin.y );
+            }
+            return distance;
+        }
+        void add( const std::vector<std::pair<point, T>> &cells ) {
+            for( const std::pair<point, T> &candidate : cells ) {
+                add( candidate );
+            }
+        }
+        void add( const std::pair<point, T> &candidate ) {
+            const int distance = rl_dist( candidate.first, range->origin );
+            if( distance > range->distance ) {
+                return;
+            }
+            sorted_elements[distance].push_back( candidate );
+        }
+        void load() {
+            const point rank_offset{ CELL_SIZE, CELL_SIZE };
+            min_bounds -= rank_offset;
+            max_bounds += rank_offset;
+            distance_to_nearest_cell_wall = calc_distance();
+            cell_ranks_loaded++;
+            const point load_offset( cell_ranks_loaded, cell_ranks_loaded );
+            const point min_cell = range->origin_cell - load_offset;
+            const point max_cell = range->origin_cell + load_offset;
+            const bool min_x_inbounds = min_cell.x >= 0;
+            const bool min_y_inbounds = min_cell.y >= 0;
+            const bool max_x_inbounds = max_cell.x < divide_round_up( MAXX, CELL_SIZE );
+            const bool max_y_inbounds = max_cell.y < divide_round_up( MAXY, CELL_SIZE );
+            // Draw a box...
+            for( int x = min_cell.x; x <= max_cell.x; ++x ) {
+                if( x < 0 || x > MAXX / CELL_SIZE ) {
+                    continue;
+                }
+                if( min_y_inbounds ) {
+                    add( range->list->cells[min_cell.y][x] );
+                }
+                if( max_y_inbounds ) {
+                    add( range->list->cells[max_cell.y][x] );
+                }
+            }
+            for( int y = min_cell.y + 1; y <= max_cell.y - 1; ++y ) {
+                if( y < 0 || y > MAXY / CELL_SIZE ) {
+                    continue;
+                }
+                if( min_x_inbounds ) {
+                    add( range->list->cells[y][min_cell.x] );
+                }
+                if( max_x_inbounds ) {
+                    add( range->list->cells[y][max_cell.x] );
+                }
+            }
+        }
+        const cell_list_range<T, CELL_SIZE, MAXX, MAXY> *range;
+        std::vector<std::vector<std::pair<point, T>>> sorted_elements;
```

Comment:
Ideally iterators should be cheap to copy.  I'm wondering how much of the data herein you could safely move to the `cell_list_range`.

Reply:
I was avoiding that because it seems more straightforward for it to be in the iterator, but now that I think about it all I have to do is shift the data...

Reply:
looks like I can move everything but this:
```
        const cell_list_range<T, CELL_SIZE, MAXX, MAXY> *range;
        int current_distance = 1;
        typename std::vector<std::pair<point, T>>::iterator current;
```

Reply:
Yes, that should be better.  Make sure you have a test where you iterate over the same range twice.

Reply:
Of course this *looked* simple, but now I have crashing and other weirdness after the refactor, so I'm retreating away from 99% done at great velocity >_<